mlir/lib/Transforms/lab4/kulaev_zhenya/KulaevZhenyaFMAPass.cpp

```diff
@@ -0,0 +1,57 @@
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+
+using namespace mlir;
+
+namespace {
+class KulaevZhenyaFMAPass
+    : public PassWrapper<KulaevZhenyaFMAPass, OperationPass<ModuleOp>> {
+public:
+  StringRef getArgument() const final { return "kulaev_zhenya_fma"; }
+  StringRef getDescription() const final {
+    return "The pass replaces the addition and multiplication operations of a "
+           "single instruction";
+  }
+
+  void runOnOperation() override {
+    ModuleOp module = getOperation();
+    module.walk([&](Operation *op) {
+      if (auto addOp = dyn_cast<LLVM::FAddOp>(op)) {
+        Value addLHS = addOp.getOperand(0);
+        Value addRHS = addOp.getOperand(1);
+
+        if (auto mulOpLHS = addLHS.getDefiningOp<LLVM::FMulOp>()) {
+          handleMulOp(addOp, mulOpLHS, addRHS);
+        } else if (auto mulOpRHS = addRHS.getDefiningOp<LLVM::FMulOp>()) {
+          handleMulOp(addOp, mulOpRHS, addLHS);
+        }
+      }
+    });
+  }
+
+private:
+  void handleMulOp(LLVM::FAddOp &addOp, LLVM::FMulOp &mulOp,
+                   Value &otherOperand) {
+    OpBuilder builder(addOp);
+    Value fma = builder.create<LLVM::FMAOp>(addOp.getLoc(), mulOp.getOperand(0),
+                                            mulOp.getOperand(1), otherOperand);
+    addOp.replaceAllUsesWith(fma);
+    addOp.erase();
+    mulOp.erase();
```

Comment:
Please, add a check that we really can remove mulOp, because it may have other uses. Also, test for that should be added

Reply:
Could you tell me in more detail exactly what you mean by the words `it may have other uses`, could you give an example, please?

Reply:
Sure, here is an example:
```
c = a * b
d = c + e
```
d is basically FMA(a,b,e) here.
But in case we have the following code

```
c = a * b
d = c + e
f = c + x
```
d is still FMA, but we are not allowed to remove `c` definition, cause it is used in calculation of `f`